src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+
+    Parameters
+    ----------
+    tag_list: list[TagReference]
+        The list of tags to check.
+    pattern: Pattern
+        The pattern to match against.
+    parts: list[str]
+        The list of version parts extracted from the version.
+    has_non_numeric_suffix : bool
+        Whether the version string has a non-numeric suffix.
+    artifact_name : str
+        The name of the artifact.
+
+    Returns
+    -------
+    list[TagReference]
+        The list of tags that matched the pattern.
+    """
+    matched_tags = []
+    for tag in tag_list:
+        tag_name = str(tag)
+        match = pattern.match(tag_name)
+        if not match:
+            continue
+        # Tags are append with their match information for possible further evaluation.
+        matched_tags.append(
+            {
+                "tag": tag,
+                "version": match.group("version"),
+                "prefix": match.group("prefix"),
+                "prefix_sep": match.group("prefix_sep"),
+                "suffix_sep": match.group("suffix_sep"),
+                "suffix": match.group("suffix"),
+            }
+        )
+
+    if len(matched_tags) <= 1:
+        return _extract_tags_from_list(matched_tags)
+
+    # In the case of multiple matches, further work must be done.
+    # Firstly, combine matches with their suffixes as some version patterns will not include the required suffix in the
+    #  version group.
+    if has_non_numeric_suffix:
+        filtered_tags = []
+        for item in matched_tags:
+            # Discard tags with no suffix or with one that does not match the version.
+            suffix: str | None = item["suffix"]
+            if not suffix:
+                filtered_tags.append(item)
+                continue
+            if suffix == parts[len(parts) - 1]:
+                filtered_tags.append(item)
+                continue
+
+        matched_tags = filtered_tags
+
+    # If any of the matches contain a prefix that matches the target artifact name, remove those that don't.
+    named_tags = []
+    for item in matched_tags:
+        prefix: str | None = item["prefix"]
+        if not prefix:
+            continue
+        if "/" in prefix:
+            # Exclude prefix parts that exists before a forward slash, e.g. rel/
+            split = prefix.split("/")
+            prefix = split[len(split) - 1]
+        if prefix.lower() == artifact_name.lower():
+            named_tags.append(item)
+
+    if named_tags:
+        matched_tags = named_tags
+
+    # If multiple tags still remain, sort them based on the closest match in terms of individual parts.
+    if len(matched_tags) > 1:
+        matched_tags.sort(
+            key=lambda matched_tag: _count_parts_in_tag(matched_tag["version"], matched_tag["suffix"], parts)
+        )
+
+    return _extract_tags_from_list(matched_tags)
+
+
+def _extract_tags_from_list(matched_tags: list[dict[str, str | TagReference]]) -> list[TagReference]:
```

Comment:
From where this function is called, shouldn't the type of the `matched_tags` param just be as follows?
```diff
-def _extract_tags_from_list(matched_tags: list[dict[str, str | TagReference]]) -> list[TagReference]:
+def _extract_tags_from_list(matched_tags: list[dict[str, str]]) -> list[TagReference]:
```
I ran this through mypy and this seems correct.

Reply:
This seems a bit strange to me as one of the items in the dictionary is a `TagReference`. Regardless, this function will be removed.